source/common/http/header_map_impl.cc

```diff
@@ -422,6 +426,9 @@ void HeaderMapImpl::addCopy(const LowerCaseString& key, absl::string_view value)
   insertByKey(std::move(new_key), std::move(new_value));
   ASSERT(new_key.empty());   // NOLINT(bugprone-use-after-move)
   ASSERT(new_value.empty()); // NOLINT(bugprone-use-after-move)
+  if (isHeadersCasePreservationEnabled()) {
```

Comment:
Sorry I am repeating myself a bit, but I think there would be a new `HeaderMapImpl::addCasePreservedCopy(const std::string& key, absl::string_view value)` (or make the key a string view, would have to check which is best based on caller)...don't change their existing methods